include/wx/webrequest.h

```diff
@@ -0,0 +1,305 @@
+#include "wx/versioninfo.h"
+
+class wxWebResponse;
+class wxWebSession;
+class wxWebAuthChallenge;
+
+WX_DECLARE_STRING_HASH_MAP(wxString, wxWebRequestHeaderMap);
+
+class WXDLLIMPEXP_NET wxWebRequest : public wxEvtHandler, public wxRefCounter
+{
+public:
+    enum State
+    {
+        State_Idle,
+        State_Unauthorized,
+        State_Active,
+        State_Completed,
+        State_Failed,
+        State_Cancelled
+    };
+
+    enum Storage
+    {
+        Storage_Memory,
+        Storage_File,
+        Storage_None
+    };
+
+    virtual ~wxWebRequest() { }
+
+    virtual void SetHeader(const wxString& name, const wxString& value)
+    { m_headers[name] = value; }
+
+    virtual void SetMethod(const wxString& method) { m_method = method; }
+
+    void SetData(const wxString& text, const wxString& contentType, const wxMBConv& conv = wxConvUTF8);
+
+    void SetData(wxSharedPtr<wxInputStream> dataStream, const wxString& contentType, wxFileOffset dataSize = wxInvalidOffset);
+
+    void SetIgnoreServerErrorStatus(bool ignore) { m_ignoreServerErrorStatus = ignore; }
+
+    virtual void SetStorage(Storage storage) { m_storage = storage; }
+
+    Storage GetStorage() const { return m_storage; }
+
+    virtual void Start() = 0;
+
+    virtual void Cancel() = 0;
+
+    virtual wxWebResponse* GetResponse() const = 0;
+
+    virtual wxWebAuthChallenge* GetAuthChallenge() const = 0;
+
+    int GetId() const { return m_id; }
+
+    wxWebSession& GetSession() const { return m_session; }
+
+    State GetState() const { return m_state; }
+
+    virtual wxFileOffset GetBytesSent() const = 0;
+
+    virtual wxFileOffset GetBytesExpectedToSend() const = 0;
+
+    virtual wxFileOffset GetBytesReceived() const;
+
+    virtual wxFileOffset GetBytesExpectedToReceive() const;
+
+    void SetState(State state, const wxString& failMsg = "");
+
+    void ReportDataReceived(size_t sizeReceived);
+
+    static void SplitParameters(const wxString& s, wxString& value,
+        wxWebRequestHeaderMap& parameters);
+
+    static void SplitParameters(const wxString::const_iterator& begin,
+        const wxString::const_iterator& end, wxWebRequestHeaderMap& parameters);
+
+protected:
+    wxString m_method;
+    Storage m_storage;
+    wxWebRequestHeaderMap m_headers;
+    wxFileOffset m_dataSize;
+    wxSharedPtr<wxInputStream> m_dataStream;
+
+    wxWebRequest(wxWebSession& session, int id):
+        m_storage(Storage_Memory),
+        m_dataSize(0),
+        m_session(session),
+        m_id(id),
+        m_state(State_Idle),
+        m_ignoreServerErrorStatus(false),
+        m_bytesReceived(0) { }
+
+    bool CheckServerStatus();
+
+    static bool IsActiveState(State state);
+
+private:
+    wxWebSession& m_session;
+    int m_id;
+    State m_state;
+    bool m_ignoreServerErrorStatus;
+    wxFileOffset m_bytesReceived;
+    wxCharBuffer m_dataText;
+
+    void ProcessStateEvent(State state, const wxString& failMsg);
+
+    wxDECLARE_NO_COPY_CLASS(wxWebRequest);
```

Comment:
I'm not sure if I'm going to be able to change this, but I think it would be much better to make `wxWebRequest` a value object containing just `wxObjectDataPtr<wxWebRequestImpl>` inside it. This would result in a simpler API at the client side, working with objects instead of pointers.

And the same thing for `wxWebResponse`.